File: libs/python/nallely_connector.py
"""
Nallely WebSocket Bus connector for Python.
LLM-generated — mirrors the behavior of libs/js/nallely-websocket.js.

Threaded connector for registering as an external neuron in a Nallely session.

Usage:

    from nallely_connector import NallelyWebsocketBus

    bus = NallelyWebsocketBus(address="192.168.1.74:6789")
    # if address is None, defaults on localhost:6789

    params = {
        "note": {"min": 0, "max": 127},
        "other_port": {"min": 0, "max": 127},
    }
    config = {"note": 0, "other_port": 0}

    service = bus.register("external", "my_neuron", params, config)
    service.onmessage = lambda msg: print(f"{msg['on']} = {msg['value']}")

    # Connection is already running at this point.
    service.send("note", 60)
    time.sleep(1)
    service.send("note", 0)

    service.send("other_port", 55.1234)

    service.dispose()
"""

import json
import struct
import threading
import time

from websockets.sync.client import connect as ws_connect


class NallelyService:
    """A single external neuron connection to the Nallely WebSocket Bus."""

    def __init__(self, kind, name, parameters, config, address=None, log=print):
        self.kind = kind
        self.name = name
        self.parameters = parameters
        self.config = config
        self.log = log

        if address:
            self.url = f"ws://{address}/{name}/autoconfig"
        else:
            self.url = f"ws://localhost:6789/{name}/autoconfig"

        self.ws = None
        self._running = False
        self._thread = None

        # Callbacks
        self.onopen = None
        self.onclose = None
        self.onerror = None
        self.onmessage = None
        self.onsend = None

    def _start(self):
        self._running = True
        self._thread = threading.Thread(target=self._run, daemon=True)
        self._thread.start()

    def _run(self):
        while self._running:
            try:
                with ws_connect(self.url) as ws:
                    self.ws = ws

                    registration = [
                        {"name": name, "range": [conf["min"], conf["max"]]}
                        for name, conf in self.parameters.items()
                    ]
                    ws.send(json.dumps({
                        "kind": self.kind,
                        "parameters": registration,
                    }))

                    if self.onopen:
                        self.onopen(registration)

                    for message in ws:
                        if isinstance(message, bytes):
                            parsed = self._parse_frame(message)
                        elif isinstance(message, str):
                            parsed = json.loads(message)
                        else:
                            continue

                        self.config[parsed["on"]] = parsed["value"]

                        if self.onmessage:
                            self.onmessage(parsed)

            except Exception as e:
                if self.onerror:
                    self.onerror(e)
                else:
                    self.log(f"Connection error: {e}")

            self.ws = None

            if self.onclose:
                self.onclose()

            if self._running:
                self.log("Reconnecting in 1 second...")
                time.sleep(1)

    def send(self, parameter, value):
        """Send a parameter value as a binary frame."""
        if self.ws:
            frame = self._build_frame(parameter, value)
            if self.onsend:
                self.onsend({"on": parameter, "value": value})
            self.ws.send(frame)
        else:
            self.log(f"WebSocket not open, cannot send: {parameter} {value}")

    def dispose(self):
        """Stop the connection loop and close the socket."""
        self._running = False
        if self.ws:
            self.ws.close()
            self.ws = None

    @staticmethod
    def _build_frame(name, value):
        name_bytes = name.encode("utf-8")
        return struct.pack("!B", len(name_bytes)) + name_bytes + struct.pack("!d", value)

    @staticmethod
    def _parse_frame(data):
        ln = data[0]
        name = data[1:1 + ln].decode("utf-8")
        value = struct.unpack_from("!d", data, 1 + ln)[0]
        return {"on": name, "value": value}


class NallelyWebsocketBus:
    """Registry of NallelyService instances."""

    def __init__(self, address=None):
        self.registered = {}
        self.address = address

    @staticmethod
    def _build_uuid(kind, name):
        return f"{kind}::{name}"

    def register(self, kind, name, parameters, config, log=print):
        """Create, register, and start a NallelyService."""
        service = NallelyService(kind, name, parameters, config, self.address, log=log)
        self.registered[self._build_uuid(kind, name)] = service
        service._start()
        return service

    def send(self, kind, name, parameter, value):
        """Send a value through a previously registered service."""
        key = self._build_uuid(kind, name)
        self.registered[key].send(parameter, value)

File: libs/ruby/nallely_connector.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
# *LLM generated*

require 'websocket-client-simple'
require 'json'

module Nallely
  # Binary frame codec for Nallely WebSocket Bus protocol
  module FrameCodec
    # Encode a parameter name and value into a binary frame
    # Format: [1 byte: name_length][N bytes: UTF-8 name][8 bytes: float64 big-endian]
    #
    # @param param_name [String] plain parameter name (e.g. "note")
    # @param value [Float] numeric value
    # @return [String] binary frame (ASCII-8BIT encoding)
    def self.encode(param_name, value)
      name_bytes = param_name.encode('UTF-8')
      name_length = name_bytes.bytesize

      raise ArgumentError, "Parameter name too long (max 255 bytes)" if name_length > 255

      frame = String.new(encoding: 'ASCII-8BIT')
      frame << [name_length].pack('C')  # uint8
      frame << name_bytes.force_encoding('ASCII-8BIT')
      frame << [value].pack('G')  # float64 big-endian (network byte order)
      frame
    end

    # Decode a binary frame into parameter name and value
    #
    # @param frame [String] binary frame
    # @return [Array<String, Float>] [param_name, value]
    def self.decode(frame)
      frame = frame.force_encoding('ASCII-8BIT')

      name_length = frame[0].unpack1('C')
      param_name = frame[1, name_length].force_encoding('UTF-8')
      value = frame[1 + name_length, 8].unpack1('G')

      [param_name, value]
    end
  end

  # External neuron service connected to the WebSocket Bus
  class ExternalNeuron
    attr_reader :name, :parameters, :connected

    # @param host [String] Nallely host (default: localhost)
    # @param port [Integer] WebSocket Bus port (default: 6789)
    # @param name [String] neuron name
    # @param parameters [Hash] parameter definitions
    #   e.g. { "note" => { min: 0, max: 127 }, "velocity" => { min: 0, max: 127 } }
    def initialize(host: 'localhost', port: 6789, name:, parameters:)
      @host = host
      @port = port
      @name = name
      @parameters = parameters
      @ws = nil
      @connected = false
      @reconnect = true
      @on_message_callback = nil
      @on_connect_callback = nil
      @on_disconnect_callback = nil
    end

    # Connect to the WebSocket Bus and register as an external neuron
    def connect
      url = "ws://#{@host}:#{@port}/#{@name}/autoconfig"

      @ws = WebSocket::Client::Simple.connect(url)

      # Capture self for use in callbacks
      neuron = self

      @ws.on :open do
        neuron.instance_variable_set(:@connected, true)
        puts "[#{neuron.name}] Connected to Nallely WebSocket Bus"
        neuron.register  # Call it directly, it's public
        neuron.instance_variable_get(:@on_connect_callback)&.call
      end

      @ws.on :message do |msg|
        neuron.send(:handle_message, msg)
      end

      @ws.on :close do |e|
        neuron.instance_variable_set(:@connected, false)
        puts "[#{neuron.name}] Disconnected: #{e}"
        neuron.instance_variable_get(:@on_disconnect_callback)&.call
        neuron.send(:auto_reconnect) if neuron.instance_variable_get(:@reconnect)
      end

      @ws.on :error do |e|
        puts "[#{neuron.name}] Error: #{e}"
      end

      # Wait for connection to be established
      sleep 0.1 until @connected

      self
    end

    # Send registration JSON to declare parameters
    def register
      registration = {
        kind: 'external',
        parameters: @parameters.map do |name, spec|
          param = { name: name }
          param[:range] = [spec[:min], spec[:max]] if spec[:min] || spec[:max]
          param
        end
      }

      @ws.send(registration.to_json)
      puts "[#{@name}] Registered with parameters: #{parameters}"
    end

    # Send a value to a parameter
    #
    # @param param_name [String] plain parameter name
    # @param value [Float] numeric value
    def send_value(param_name, value)
      unless @connected
        puts "[#{@name}] Warning: Not connected, cannot send #{param_name}=#{value}"
        return
      end

      frame = FrameCodec.encode(param_name, value.to_f)
      @ws.send(frame, type: :binary)
    end

    # Set callback for incoming messages
    #
    # @yield [param_name, value] called when a value is received
    def on_message(&block)
      @on_message_callback = block
    end

    # Set callback for connection established
    #
    # @yield called when connected
    def on_connect(&block)
      @on_connect_callback = block
    end

    # Set callback for disconnection
    #
    # @yield called when disconnected
    def on_disconnect(&block)
      @on_disconnect_callback = block
    end

    # Disconnect cleanly
    def disconnect
      @reconnect = false
      @ws&.close
      @connected = false
    end

    # Unregister from the session (removes parameters and wiring)
    def unregister
      disconnect
      # Connect to unregister endpoint
      url = "ws://#{@host}:#{@port}/#{@name}/unregister"
      unregister_ws = WebSocket::Client::Simple.connect(url)
      sleep 0.5  # Give it time to process
      unregister_ws.close
      puts "[#{@name}] Unregistered"
    end

    private

    # Handle incoming WebSocket message
    def handle_message(msg)
      if msg.type == :binary
        # Binary frame with parameter value
        param_name, value = FrameCodec.decode(msg.data)
        @on_message_callback&.call(param_name, value)
      elsif msg.type == :text
        # JSON message format: {"on": "param_name", "value": float}
        data = JSON.parse(msg.data)
        if data['on'] && data['value']
          @on_message_callback&.call(data['on'], data['value'])
        end
      end
    rescue => e
      puts "[#{@name}] Error handling message: #{e.message}"
    end

    # Auto-reconnect after disconnect
    def auto_reconnect
      return unless @reconnect

      puts "[#{@name}] Reconnecting in 1 second..."
      sleep 1
      connect
    rescue => e
      puts "[#{@name}] Reconnection failed: #{e.message}"
      auto_reconnect
    end
  end

  # Service registry for managing multiple external neurons
  class WebSocketBus
    def initialize(host: 'localhost', port: 6789)
      @host = host
      @port = port
      @services = {}
    end

    # Register a new external neuron
    #
    # @param name [String] neuron name
    # @param parameters [Hash] parameter definitions
    # @return [ExternalNeuron] the registered neuron
    def register(name, parameters)
      service_key = "external::#{name}"

      if @services[service_key]
        puts "Service #{service_key} already registered"
        return @services[service_key]
      end

      neuron = ExternalNeuron.new(
        host: @host,
        port: @port,
        name: name,
        parameters: parameters
      )

      neuron.connect
      @services[service_key] = neuron
      neuron
    end

    # Get a registered service
    #
    # @param name [String] neuron name
    # @return [ExternalNeuron, nil]
    def get(name)
      @services["external::#{name}"]
    end

    # Send a value through a registered service
    #
    # @param name [String] neuron name
    # @param param_name [String] parameter name
    # @param value [Float] value to send
    def send_value(name, param_name, value)
      neuron = get(name)
      if neuron
        neuron.send_value(param_name, value)
      else
        puts "Service #{name} not found"
      end
    end

    # Disconnect all services
    def disconnect_all
      @services.each_value(&:disconnect)
      @services.clear
    end

    # Unregister all services
    def unregister_all
      @services.each_value(&:unregister)
      @services.clear
    end
  end
end

# Example usage
if __FILE__ == $PROGRAM_NAME
  # Create a simple external neuron with input and output parameters
  neuron = Nallely::ExternalNeuron.new(
    host: 'localhost',
    port: 6789,
    name: 'ruby_test',
    parameters: {
      'input' => { min: 0, max: 127 },
      'output' => { min: 0, max: 127 }
    }
  )

  # Set up message handler
  neuron.on_message do |param_name, value|
    puts "Received: #{param_name} = #{value}"

    # Echo back with modification
    if param_name == 'input'
      neuron.send_value('output', value * 0.5)
    end
  end

  # Connect
  neuron.connect

  # Send test values
  neuron.on_connect do
    puts "Sending test values..."
    10.times do |i|
      neuron.send_value('output', i * 12.7)
      sleep 0.5
    end
  end

  # Keep the script running
  puts "External neuron running. Press Ctrl+C to exit."
  begin
    sleep
  rescue Interrupt
    puts "\nShutting down..."
    neuron.disconnect
  end
end
